src/tools/compiletest/src/header.rs

```diff
@@ -963,6 +963,21 @@ pub fn make_test_description<R: Read>(
         .join(if config.host.contains("windows") { "rust-lld.exe" } else { "rust-lld" })
         .exists();
 
+    let path_env = env::var_os("PATH").unwrap();
+    let split_path_env = env::split_paths(&path_env).collect::<Vec<_>>();
+    let is_on_path = |file| split_path_env.iter().any(|dir| dir.join(file).is_file());
+
+    // On Windows, dlltool.exe is used for all architectures.
+    #[cfg(windows)]
+    let (has_i686_dlltool, has_x86_64_dlltool) = {
+        let has_dlltool = is_on_path("dlltool.exe");
+        (has_dlltool, has_dlltool)
+    };
+    // For non-Windows, there are architecture specific dlltool binaries.
+    #[cfg(not(windows))]
+    let (has_i686_dlltool, has_x86_64_dlltool) =
+        (is_on_path("i686-w64-mingw32-dlltool"), is_on_path("x86_64-w64-mingw32-dlltool"));
```

Comment:
These look like an expensive checks to do for every test case. Can you make them lazy by wrapping in a closure and only invoke it if a `needs-*-dlltool` directive is actually encountered? The `&&` operator is short-circuiting, so switching the terms inside `reason!()` should accomplish that.

Reply:
Done